docs/design/resource-qos.md

```diff
@@ -0,0 +1,224 @@
+[here](http://releases.k8s.io/release-1.0/docs/design/resource-qos.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Resource Quality of Service in Kubernetes
+
+**Author(s)**: Ananya Kumar (@AnanyaKumar), Vishnu Kannan (vishh@)
+**Last Updated**: 10/5/2015
+
+**Status**: Implemented
+
+*This document presents the design of resource quality of service for containers in Kubernetes, and describes use cases and implementation details.*
+
+## Motivation
+
+In the first version of Kubernetes, users specified one value, a resource limit, for each resource of a container.
+For example, a user might specify 1GB memory limit.
+The scheduler used resource *limits* to schedule containers (technically, the scheduler schedules pods comprised of containers).
+For example, the scheduler would not place 5 containers each with a 1GB memory limit onto a machine with 4GB of memory.
+
+This had the drawback that if users over-estimate the resource needs of their containers, then cluster utilization will be poor, since the unused resources are stranded and can't be used by other containers.
+On the other hand, if they under-estimate the resource needs (specifying no limits is a special case of under-estimating), then the machine could become overcommitted, leading to out-of-memory or CPU starvation of any container on the node, even the ones that specified accurate resource limits.
+
+This document describes the way Kubernetes provides different levels of Quality of Service to containers depending on what they *request*.
+Containers that need to stay up reliably can request guaranteed resources, while containers with less stringent requirements can use resources with weaker or no guarantee.
+Specifically, for each resource, containers specify a request, which is the amount of that resource that the system will guarantee to the container, and a limit which is the maximum amount that the system will allow the container to use.
+When request == limit, the resources are guranteed, and when request < limit, the container is guaranteed the request but can opportunistically scavenge the difference between request and limit if they are not being used by other containers.
+This allows Kubernetes to oversubscribe nodes, which increases utilization, while at the same time maintaining resource guarantees for the containers that need guarantees.
+Borg increased utilization by about 20% when it started allowing use of such non-guaranteed resources, and we hope to see similar improvements in Kubernetes.
+
+## Requests and Limits
+
+For each resource, containers can specify a resource request and limit, 0 <= request <= limit <= Infinity.
+If the container is successfully scheduled, the container is guaranteed the amount of resource requested.
+Scheduling is based on `requests` and not `limits`.
```

Comment:
I think it's important to note what the defaulting logic is so there is no confusion when users specify limits and not requests.  If a limit is specified, but there is no corresponding request, the request defaults to the limit.